File: typescript/Behavioral/19-State/State/PendingReviewState.ts
import { BaseJobState } from "./BaseJobState";
import { ClosedState } from "./ClosedState";
import { InProgressState } from "./InProgressState";
import type { JobCard } from "../Context/JobCard";

export class PendingReviewState extends BaseJobState {
  readonly name = "PendingReview";

  override approve(job: JobCard): void {
    job.closedAt = new Date();
    job.transitionTo(new ClosedState());
    console.log("  🔒 Approved and closed");
  }

  // Review rejected — send back to InProgress
  override start(job: JobCard): void {
    job.completedAt = undefined;
    job.transitionTo(new InProgressState());
    console.log("  ↩️  Review rejected — reopened for rework");
  }
}

File: csharp/Behavioral/19-State/State/PendingReviewState.cs
namespace State;

public class PendingReviewState : BaseJobState
{
    public override string Name => "PendingReview";

    public override void Approve(JobCard job)
    {
        job.ClosedAt = DateTime.Now;
        job.TransitionTo(new ClosedState());
        Console.WriteLine("  🔒 Approved and closed");
    }

    // Review rejected — send back to InProgress
    public override void Start(JobCard job)
    {
        job.CompletedAt = null;
        job.TransitionTo(new InProgressState());
        Console.WriteLine("  ↩️  Review rejected — reopened for rework");
    }
}
